### PR TITLE
fix 3d optimizers for grouped convolutions and add more test cases

### PR DESCRIPTION
**Pull request by affanv14**
addresses https://github.com/Theano/Theano/issues/6360

### CODE REVIEW

**Comment by affanv14** [comment]
Currently, the check if an optimizer returns None is based on the excluding conv_gemm and conv_dnn and also ensuring that the meta-optimizer returns None which causes an error. 
I feel this is too roundabout and I will improve that aspect.

The tests are getting bloated though, is it alright to have so mant test cases?

**Comment by gvtulder** [comment]
Thanks for the fast work on this. You could also try to test the individual opts that aren't tested elsewhere, instead of using the metaoptimizer for all of them. The metaoptimizer itself is not affected by the groups setting, so it should be sufficient to test the optimizers directly. That might be a bit faster and it would be easier to check if the optimization returns None.

Once the groups test work, is it possible to add `unshared = True` to these tests, while you're at it?

**Comment by affanv14** [comment]
> You could also try to test the individual opts that aren't tested elsewhere, instead of using the metaoptimizer for all of them

The individual opts cannot be tested since the are many optimizations which are applied before that( one of them is transferring to the gpu).
Currently, I see two ways of doing this ( other suggestions are welcome!). One is the method I implemented above, the other is to make sure the op which is used in the meta-optimizer is not the same as the op used by the function after optimization.

I have a third method which just involves converting the theano tensor into a gpuarray and then apply the individual opts, but I don't know if this will work.  So, I'll try this method now.

**Comment by affanv14** [comment]
The third method I mentioned seems too volatile.

**Comment by nouiz** [comment]
Why do you think the 3rd option is too volatile? This is what we do elsewhere and it go well.

**Comment by nouiz** [comment]
jenkins test this

**Comment by affanv14** [comment]
> Why do you think the 3rd option is too volatile

Since jenkins uses `init_gpu_device `and acts like it is in the cpu unless `mode_with_gpu `is passed to the function, the tensors initialized are not of type gpuarray. I tried to convert these to type `gpuarray `using `as_gpuarray_variable `after which I passed these as input to `AbstractConv2d`. However the `AbstractConv2d  `converts the inputs back using `hostfromgpu`. I thought if I dive into the specifics here and try to traverse the graph, it would go against operations when `device=cuda0` is given(since here shared variable are already in gpuarray). So the volatility I am assuming is when `init_gpu_device `is used and `device=cuda0`(or any gpu device) is used

Am I missing something?

**Comment by nouiz** [comment]
If you pass a GpuArrayVariable() instance to AbstractConv2d(), it should
not move it to the GPU. This is what the optimization that move it to the
GPU does. I'll try to find you some code that does that as an example.

**Comment by affanv14** [comment]
> If you pass a GpuArrayVariable() instance to AbstractConv2d(), it should
not move it to the GPU

Sorry I did not explain completely, after passing the GpuArrayVariable() instance to AbstractConv2d() I take the abstractconv2d node and  pass it to  one of the optimization functions which needs to be tested. but the optimizers return None because its immediate inputs are not of gpuarray type. I think this is due to AbstractConv2d using hostfromgpu to convert it back to tensortype.

>  I'll try to find you some code that does that as an example.

yes, that would be great

**Comment by nouiz** [comment]
Here is a diff that help does the 3rd case.

When jenkins pass, I'll merge this PR as we wan to make a beta2 shortly and
this PR contain fixes. The faster test can be done in another PR.

thanks

```
diff --git a/theano/gpuarray/tests/test_dnn.py
b/theano/gpuarray/tests/test_dnn.py
index e280504df..5b96e5b81 100644
--- a/theano/gpuarray/tests/test_dnn.py
+++ b/theano/gpuarray/tests/test_dnn.py
@@ -18,7 +18,7 @@ from theano.tensor.nnet.abstract_conv import
get_conv_output_shape, get_conv_gra
 from theano.tensor.nnet import bn

 from .. import dnn
-from ..basic_ops import GpuAllocEmpty
+from ..basic_ops import as_gpuarray_variable, GpuAllocEmpty
 from ..type import gpuarray_shared_constructor, GpuArrayType

 from .config import mode_with_gpu, mode_without_gpu, test_ctx_name,
ref_cast
@@ -2723,3 +2723,13 @@ def test_conv_guess_once_with_dtypes():
     f_pseudo_half_config()
     f_float_config()
     f_double_config()
+
+
+def test_gpu_abstract_conv():
+    i = theano.tensor.tensor4()
+    f = theano.tensor.tensor4()
+    gi = as_gpuarray_variable(i, test_ctx_name)
+    gf = as_gpuarray_variable(f, test_ctx_name)
+    o = theano.tensor.nnet.abstract_conv.AbstractConv2d()(gi, gf)
+    theano.printing.debugprint(o)
+    assert theano.gpuarray.dnn.local_abstractconv_cudnn.transform(o.owner)
```

**Comment by affanv14** [comment]
Alright

**Comment by affanv14** [comment]
Thanks for the example.

File: theano/gpuarray/dnn.py [hunk 1]
```diff
@@ -3274,7 +3275,7 @@ def local_abstractconv3d_cudnn_alt(node):
         conv_mode = 'cross'
 
     if isinstance(op, AbstractConv3d):
-        if border_mode == 'half' or subsample != (1, 1, 1):
+        if border_mode == 'half' or subsample != (1, 1, 1) or num_groups > 1:
```
[the remark this answers is not in the document]
Comment (affanv14) [reply theano/gpuarray/dnn.py]: Unshared is only for 2d optimizers, these fixes are for 3d optimizers. so it is not needed..

Only unshared test cases need to be added, which I planned to do in the new PR

**Comment by nouiz** [comment]
thanks

**Pull request merged**